docs/docs/internals/explicit-nulls.md

```diff
@@ -0,0 +1,182 @@
+          `x != null && x.length < 10`, return the `Inferred` facts.
+           
+ 	  In turn, `Inferred` is defined as `case class Inferred(ifTrue: FlowFacts, ifFalse: FlowFacts)`.
+          That is, `Inferred` contains the paths that _must_ be non-null if the condition is true and,
+          separately, the paths that must be non-null if the condition is false.
+
+          e.g. for `x != null` we'd get `Inferred({x}, {})`, but only if `x` is stable.
+          However, if we had `x == null` we'd get `Inferred({}, {x})`.
+      	  
+      2. `inferWithinCond(cond: Tree): FlowFacts`: given a condition of the form `lhs && rhs` or
+         `lhs || rhs`, calculate the paths that must be non-null for the rhs to execute (given
+         that these operations) are short-circuiting.
+
+      3. `inferWithinBlock(stat: Tree): FlowFacts`: if `stat` is a statement with a block, calculate
+         which paths must be non-null when the statement that _follows_ `stat` in the block executes.
+         This is so we can handle things like
+         ```scala
+         val x: String|Null = ???
+	 if (x == null) return
+         val y = x.length
+         ```
+         Here, `inferWithinBlock(if (x == null) return)` gives back `{x}`, because we can tell that
+         the next statement will execute only if `x` is non-null.
+
+      4. `refineType(tpe: Type): Type`: given a type, refine it if possible using flow-sensitive type
+         information. This uses a `NonNullTermRef` (see below).
+
+  - Each of the public APIs in `FlowTyper` is used to do flow typing in a different scenario
+    (but all the use sites of `FlowTyper` are in `Typer.scala`):
+    * `refineType` is used in `typedIdent` and `typedSelect`
+    * `inferFromCond` is used for typing if statements
+    * `inferWithinCond` is used when typing "applications" (which is how "&&" and "||" are encoded)
+    * `inferWithinBlock` is used when typing blocks
+
+    For example, to do FlowTyping on if expressions:
+      * we type the condition
+      * we give the typed condition to the FlowTyper and obtain a pair of sets of paths `(ifTrue, ifFalse)`.
+        We type the `then` branch with the `ifTrue` facts, and the else branch with the `ifFalse` facts.
+      * profit
+
+Flow typing also introduces two new abstractions: `NonNullTermRef` and `ValDefInBlockCompleter`.
+
+#### NonNullTermRef
+This is a new type of `TermRef` (path-dependent type) that, whenever its denotation is updated, makes sure
+that the underlying widened type is non-null. It's defined in `Types.scala`. A `NonNullTermRef` is identified by `computeDenot` whenever the denotation is updated, and then we call `stripNull` on the widened type.
+
+To use the flow-typing information, whenever we see a path that we know must be non-null (in `typedIdent` or
+`typedSelect`), we replace its `TermRef` by a `NonNullTermRef`.
+
+#### ValDefInBlockCompleter
```

Comment:
I believe this is not correct. Completers have to run in the creation context, never in the completion context. Otherwise you will get subtle errors like that a completer is run later from a nested context, where symbols are visible that should not be visible at the completion context. It's an iron principle that completers are always run with the creation context.

On the other hand, the use case why we need this does not seem overly important. Returns in blocks are rare. So we really need to cater for them?

Reply:
On second thought, we might get away with this. The question is, when is a completer for a ValDef in a block called out of order? There could be a reference to the ValDef before its definition, like this:
```scala
val x = y + 1
val y = ...
```
But that would be an error that's detected later in `RefChecks. The other situation I see right now is if `y` is implicit and an implicit is searched earlier. Maybe `y`'s type does not match but it is completed anyway. E.g.
```scala
def x = implicitly[C]
import foo.bar
implicit val c: C = ... bar ...
```
If we use the creation context for `c`, we will not see `bar`.

One possibility I see is to have a completion context that uses the calling context IF it is forced in sequence in `typedStats`, and the creation context otherwise.

Reply:
Looking at the current implementation, we only use the new kind of completer (the one that uses the completion context) when the symbol is a `val` that't not lazy or implicit. So your second example shouldn't be a problem. The first example is still incorrectly typecheck with the wrong context, but like you mentioned it gets caught in refchecks. Is this an acceptable compromise?

Reply:
I am still not convinced. What are the reasons not to use the new Completer for lazy or implicit vals? It smells like an unhealthy interactions between concepts that should stay separate. I think it's safer to use the following scheme:

In `Namer#Completer`, change `completeInCreationContext` from
```
    def completeInCreationContext(denot: SymDenotation): Unit = ...
```
to
```
    def completeWithContext(denot: SymDenotation, ctx: Context): Unit = ...
```
Change the call of `completeInCreationContext` in `complete` to
```
        completeWithContext(denot, ctx)
```
Then, in `Typer#typedStats#traverse` you see:
```
      case (mdef: untpd.DefTree) :: rest =>
        mdef.removeAttachment(ExpandedTree) match {
          case Some(xtree) =>
            traverse(xtree :: rest)
          case none =>
            ...
```
At this point insert something like:
```
    mdef.symbol.infoOrCompleter match {
       case completer: Namer#Completer =>
         completer.completeWithContext(mdef.symbol, ctx)
      case _ =>
    }
```
I believe that should work. The idea is that we use creation context whenever we need to force a completer from some different context, but use calling context if we get to a definition in a statement sequence that has not been completed yet. Note that in this case the position of calling context and creation context coincide, but calling context knows about more things than creation context. So that looks safe.

Reply:
Done. We implemented this.